### PR TITLE
Remove `RateLimitedError` and make the global ratelimit abide by `max_rate_limit`

### LINKED ISSUE #746: Remove RateLimitedError
### Summary
RateLimitedError is unnecessary and could sanely be replaced by per-request retries or raising RateLimitTooLongError if the retry-after is larger than the set max.

### Problem
Since RateLimitTooLongError was added, the need for RateLimitedError has been essentially replaced by it. Having multiple ratelimited errors can be annoying especially when a lot of out-of-bucket ratelimits which currently lead to a RateLimitedError are short enough that it'd be sane to automatically retry after that time unless the retry-after time is greater than the set max for retry after.

### Ideal implementation
I propose that RateLimitedError be replaced by a per-request automatic-retry after a sleep if the retry-after is under the defined max otherwise RateLimitTooLongError will be raised

### Checklist

- [x] I have searched the issue tracker and have made sure it's not a duplicate.
      If it is a follow up of another issue, I have specified it.

### LINKED ISSUE #840: Global ratelimit handling isn't respecting max_rate_limit
This line relies on a ManualRateLimiter instance which is completely un-aware of max_rate_limit for backing off.

https://github.com/hikari-py/hikari/blob/b33a133c0214970355a2a6afb956a45fd38bfc31/hikari/impl/rest.py#L903-L909

This line which pre-empts global rate-limits also relies on the same ManualRateLimiter instance which is unaware of max_rate_limit

https://github.com/hikari-py/hikari/blob/b33a133c0214970355a2a6afb956a45fd38bfc31/hikari/impl/rest.py#L730

For reference the global rate limit handler is initialised in the following code 
https://github.com/hikari-py/hikari/blob/b33a133c0214970355a2a6afb956a45fd38bfc31/hikari/impl/rest.py#L466-L469

### Ideal Solution

These lines should both raise RateLimitedTooLong if max_rate_limit is violated by the global ratelimit.

### PR DESCRIPTION
**Pull request by davfsa**
### Checklist

- [x] I have run `nox` and all the pipelines have passed.
- [x] I have made unittests according to the code I have added/modified/deleted.

### Related issues
Closes #746 
Closes #840

### CODE REVIEW

File: hikari/errors.py [hunk 1]
```diff
@@ -411,20 +380,21 @@ class RateLimitTooLongError(HTTPError):
     reset_at: float = attr.field()
     """UNIX timestamp of when this limit will be lifted."""
 
-    limit: int = attr.field()
-    """The maximum number of calls per window for this rate limit."""
+    limit: undefined.UndefinedOr[float] = attr.field()
```
Comment (FasterSpeeding) [inline hikari/errors.py:383]: why not just `Optional[float]` for both of these

File: hikari/impl/buckets.py [hunk 2]
```diff
@@ -383,6 +399,10 @@ def __init__(self, max_rate_limit: float) -> None:
         self._max_rate_limit = max_rate_limit
         self._global_ratelimit = rate_limits.ManualRateLimiter()
 
+    @property
+    def max_rate_limit(self) -> float:
+        return self._max_rate_limit
```
Comment (FasterSpeeding) [inline hikari/impl/buckets.py:404]: missing test case

File: hikari/errors.py (hunk 1 above)
Comment (davfsa) [inline hikari/errors.py:383]:
```suggestion
    limit: typing.Optional[float] = attr.field()
```

File: hikari/errors.py [hunk 3]
```diff
@@ -55,6 +54,7 @@
 
 import attr
 
+from hikari import undefined
```
Comment (davfsa) [inline hikari/errors.py:57]:
```suggestion
```

File: hikari/errors.py [hunk 4]
```diff
@@ -411,20 +380,21 @@ class RateLimitTooLongError(HTTPError):
...
     """UNIX timestamp of when this limit will be lifted."""
 
-    limit: int = attr.field()
-    """The maximum number of calls per window for this rate limit."""
+    limit: undefined.UndefinedOr[float] = attr.field()
+    """The maximum number of calls per window for this rate limit, if known."""
 
-    period: float = attr.field()
-    """How long the rate limit window lasts for from start to end."""
+    period: undefined.UndefinedOr[float] = attr.field()
```
Comment (davfsa) [inline hikari/errors.py:386]:
```suggestion
    period: typing.Optional[float] = attr.field()
```

File: hikari/impl/buckets.py [hunk 5]
```diff
@@ -302,13 +304,27 @@ async def acquire(self) -> None:
...
+            # Release lock before we error
+            self._lock.release()
+            raise errors.RateLimitTooLongError(
+                route=self._compiled_route,
+                is_global=True,
+                retry_after=global_ratelimit.reset_at - now,
+                max_retry_after=self._max_rate_limit,
+                reset_at=global_ratelimit.reset_at,
+                limit=undefined.UNDEFINED,
+                period=undefined.UNDEFINED,
```
Comment (davfsa) [inline hikari/impl/buckets.py:318]:
```suggestion
                limit=None,
                period=None,
```

File: hikari/impl/buckets.py [hunk 6]
```diff
@@ -191,6 +191,7 @@
 import typing
 
 from hikari import errors
+from hikari import undefined
```
Comment (davfsa) [inline hikari/impl/buckets.py:194]:
```suggestion
```

File: hikari/impl/rest.py [hunk 7]
```diff
@@ -938,23 +942,28 @@ async def _parse_ratelimits(
...
-
-        raise errors.RateLimitedError(
-            url=str(response.real_url),
-            route=compiled_route,
-            headers=response.headers,
-            raw_body=body,
-            retry_after=body_retry_after,
+            return 0
+
+        _LOGGER.error(
```
Comment (davfsa) [inline hikari/impl/rest.py:947]: Add `resouce` bucket header here for more info

File: hikari/impl/rest.py [hunk 8]
```diff
@@ -938,23 +942,28 @@ async def _parse_ratelimits(
...
 
+        if body_retry_after > self._bucket_manager.max_rate_limit:
+            raise errors.RateLimitTooLongError(
+                route=compiled_route,
+                is_global=False,
+                retry_after=body_retry_after,
+                max_retry_after=self._bucket_manager.max_rate_limit,
+                reset_at=time.monotonic() + body_retry_after,
+                limit=undefined.UNDEFINED,
+                period=undefined.UNDEFINED,
```
Comment (davfsa) [inline hikari/impl/rest.py:962]:
```suggestion
                limit=None,
                period=None,
```

File: hikari/errors.py [hunk 9]
```diff
@@ -411,20 +379,21 @@ class RateLimitTooLongError(HTTPError):
     reset_at: float = attr.field()
     """UNIX timestamp of when this limit will be lifted."""
 
-    limit: int = attr.field()
-    """The maximum number of calls per window for this rate limit."""
+    limit: typing.Optional[float] = attr.field()
```
Comment (davfsa) [inline hikari/errors.py:382]: I think float is a copy issue here. Should be int

File: hikari/impl/buckets.py (hunk 2 above)
Comment (davfsa) [reply hikari/impl/buckets.py:403]: Added :)

**Review by FasterSpeeding — APPROVED** [review]

**Pull request merged**